File: Applied Course/4.Problem Solving/14.Maths & Bit Manipulation/5.Total Hamming Distance.cpp
#include<bits/stdc++.h>
using namespace std;
// Logic-1
class Solution {
public:
     int count1bits(int n)
     {
        int count=0;
        while(n>0)
        {
            n&=n-1;
            count+=1;
        }
        return count;
     }
    int totalHammingDistance(vector<int>& nums) {
        int result=0;
        for(int i=0;i<nums.size()-1;i++)
            for(int j=i+1;j<nums.size();j++)
                result+=count1bits(nums[i]^nums[j]);
        return result;
    }
};
// Logic-2
class Solution {
public:
    int totalHammingDistance(vector<int>& nums) {
        int n=nums.size();
        int result=0;
        for(int i=0;i<32;i++)
        {
            int count=0;
            for(int j=0;j<n;j++)
                if(nums[j]&(1<<i))
                    count+=1;
            result+=(count*(n-count));
        }
        return result;
    }
};
int main()
{
    // https://leetcode.com/problems/total-hamming-distance
    return 0;
}

File: Applied Course/4.Problem Solving/14.Maths & Bit Manipulation/5.Total Hamming Distance.py
# https://leetcode.com/problems/total-hamming-distance
# Logic-1
class Solution:
    def totalHammingDistance(self, nums: List[int]) -> int:
        def count1bits(n):
            count=0
            while n>0:
                n&=n-1
                count+=1
            return count
        result=0
        for i in range(len(nums)-1):
            for j in range(i+1,len(nums)):
                result+=count1bits(nums[i]^nums[j])
        return result
# Logic-2
class Solution:
    def totalHammingDistance(self, nums: List[int]) -> int:
        n=len(nums)
        result=0
        for i in range(32):
            count=0
            for j in range(n):
                if nums[j]&(1<<i):
                    count+=1
            result+=(count*(n-count))
        return result;